tools/shared/src/main/scala/org/scalajs/core/tools/optimizer/OptimizerCore.scala

```diff
@@ -2168,6 +2168,112 @@ private[optimizer] abstract class OptimizerCore(
     }
   }
 
+  /** Translate literals to their ScalaJS String representation. */
+  private def foldToStringForString_+(tree: Tree)(implicit pos : Position): Tree = tree match {
+    case FloatLiteral(value)   => StringLiteral(jsNumberToString(value.toDouble))
+    case DoubleLiteral(value)  => StringLiteral(jsNumberToString(value))
+    case LongLiteral(value)    => StringLiteral(value.toString)
+    case IntLiteral(value)     => StringLiteral(value.toString)
+    case BooleanLiteral(value) => StringLiteral(value.toString)
+    case Null()                => StringLiteral("null")
+    case Undefined()           => StringLiteral("undefined")
+    case _                     => tree
+  }
+
+  /* Following the ECMAScript 6 specification */
+  private def jsNumberToString(value: Double): String = {
+    @inline def getJVMnks: (Int, Int, BigInt) = {
+      
+      // The JVM's to String works in most cases
+      val jvmString = value.toString
+      val n = math.floor(math.log10(value)).toInt + 1
+      
+      val sString = jvmString.filter(_ != '.').takeWhile(_.isDigit)
+
+      // The JVM does love its .0
+      val corrected = sString.stripSuffix("0")
+      
+      val s = BigInt(corrected)
+      val k = corrected.length
+
+      val bits = java.lang.Double.doubleToLongBits(value)
+      val offsetExp = (bits & 0x7fff000000000000L) >> 52
+      val exp = (offsetExp - 1023).toInt
+
+      // Semantics are different only when ulp > 1 (exp > 53)
+      // And js doesn't use scientific notation (n <= 21)
+      if(exp >= 53 && n <= 21){
+       
+        @inline def toOriginalValue(s: BigInt): BigInt = 
+          s * BigInt(10).pow(n - s.toString.length)
+
+        @inline def dropTrailingZeros(s: BigInt): BigInt = 
+          BigInt(s.toString.reverse.dropWhile(_ == '0').reverse)
+
+        val shorterS = (0 to (k - 12)).flatMap( i => 
+          Seq(
+            BigDecimal(s).setScale(-i, BigDecimal.RoundingMode.UP).toBigInt,  
+            BigDecimal(s).setScale(-i, BigDecimal.RoundingMode.DOWN).toBigInt
+          )
+        ).filter(
+          s => toOriginalValue(s).toDouble == value  
+        ).map(
+          s => dropTrailingZeros(s)
+        ).distinct
+        
+        val newK = shorterS.map(_.toString.length).min
+
+        val shortestS = shorterS.filter(_.toString.length == newK)
+        
+        val mantissa = (bits & 0x000fffffffffffffL) + (1L << 52) 
+        val exactValue = BigInt(mantissa) << (exp - 52)
+
+        @inline def distance(s: BigInt): BigInt =
+          (toOriginalValue(s) - exactValue).abs
+
+        val minDistance = shortestS.map(s => distance(s)).min
+
+        val closestS = shortestS.filter(s => distance(s) == minDistance)
+
+        val newS = if(closestS.length == 1) closestS.head else closestS.filter(_ % 2 == 0).head
+
+        (n, newK, newS)
```

Comment:
Yes, but then again, is it worth underspecifying `toString` for the benefit of being to constant-fold? I don't think so either.